chainio/txmgr/geometric/geometric.go

```diff
@@ -0,0 +1,598 @@
+	retryFromFailure := 0
+	from, err := t.wallet.SenderAddress(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get sender address: %w", err)
+	}
+	for retryFromFailure < t.params.MaxSendTransactionRetry {
+		gasTipCap, err := t.estimateGasTipCap(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to estimate gas tip cap: %w", err)
+		}
+		txn, err = t.updateGasTipCap(ctx, req.tx, gasTipCap, from)
+		if err != nil {
+			return nil, fmt.Errorf("failed to update gas price: %w", err)
+		}
+		txID, err = t.wallet.SendTransaction(ctx, txn)
+		// TODO: what is this...? does it come from the fireblocks wallet?
+		var timeout interface{ Timeout() bool }
+		if errors.As(err, &timeout); timeout != nil && timeout.Timeout() {
+			t.logger.Warn(
+				"failed to send txn due to timeout",
+				"hash",
+				txn.Hash().Hex(),
+				"numRetries",
+				retryFromFailure,
+				"maxRetry",
+				t.params.MaxSendTransactionRetry,
+				"err",
+				err,
+			)
+			// TODO: why do we only retry on client or server timeouts? what about other errors?
+			retryFromFailure++
+			continue
+		} else if err != nil {
+			return nil, fmt.Errorf("failed to send txn %s: %w", txn.Hash().Hex(), err)
+		} else {
+			t.logger.Debug("successfully sent txn", "txID", txID, "txHash", txn.Hash().Hex())
+			break
+		}
+	}
+
+	if txn == nil || txID == "" {
+		return nil, fmt.Errorf("failed to send txn %s: %w", req.tx.Hash().Hex(), err)
+	}
+
+	req.tx = txn
+	req.txAttempts = append(req.txAttempts, &transaction{
+		TxID:        txID,
+		Transaction: txn,
+		requestedAt: time.Now(),
+	})
+
+	receipt, err := t.monitorTransaction(ctx, req)
+	if err == nil {
+		if receipt.GasUsed > 0 {
+			t.metrics.ObserveGasUsedWei(receipt.GasUsed)
+		}
+		t.metrics.ObserveConfirmationLatencyMs(time.Since(req.requestedAt).Milliseconds())
+	}
+	return receipt, err
+}
+
+// ensureAnyTransactionBroadcasted waits until all given transactions are broadcasted to the network.
+func (t *GeometricTxManager) ensureAnyTransactionBroadcasted(ctx context.Context, txs []*transaction) error {
+	queryTicker := time.NewTicker(t.params.GetTxReceiptTickerDuration)
+	defer queryTicker.Stop()
+
+	for {
+		for _, tx := range txs {
+			_, err := t.wallet.GetTransactionReceipt(ctx, tx.TxID)
+			if err == nil || errors.Is(err, ethereum.NotFound) || errors.Is(err, wallet.ErrReceiptNotYetAvailable) {
+				t.metrics.ObserveBroadcastLatencyMs(time.Since(tx.requestedAt).Milliseconds())
+				return nil
+			}
+		}
+
+		// Wait for the next round.
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-queryTicker.C:
+		}
+	}
+}
+
+func (t *GeometricTxManager) ensureAnyTransactionConfirmed(
+	ctx context.Context,
+	txs []*transaction,
+) (*types.Receipt, error) {
+	queryTicker := time.NewTicker(t.params.GetTxReceiptTickerDuration)
+	defer queryTicker.Stop()
+	var receipt *types.Receipt
+	var err error
+	// transactions that need to be queried. Some transactions will be removed from this map depending on their status.
```

Comment:
wdym by `some txs will be removed` and where is it removed??

Reply:
oh I see the `delete` in line 323

Reply:
In Java you are never allowed to remove an element while iterating but Golang allows it - interesting - I hope it does not affect iteration lol

Reply:
That's a good question... only found https://groups.google.com/g/golang-nuts/c/rEmaoxi11_A which seems to be say that it's fine... but it's super old. and the official go spec and faq don't seem to mention this anywhere. :\

Reply:
[bb4ae1f](https://github.com/Layr-Labs/eigensdk-go/pull/284/commits/bb4ae1ff3eca55564fbcce796820ec411f109121)
added comment here. maybe some other kind future soul will catch this and know the correct reference